### ISSUE TITLE
GPL License

### ISSUE THREAD

**Issue by lmm-git** [rcatolino/pam_sm_rust#33]
This project is currently licensed under GPL 3.0, which is a great license!

However, in my experience this is quite an unusual choice for such a library in the rust ecosystem as it essentially means (I believe) that [all binaries/crates using the library must/will be licensed under GPL3 as well](https://www.reddit.com/r/rust/comments/jk5g20/eli5_why_do_so_less_people_use_agplv3_for_their/).

This would be true e.g. for [shavee](https://github.com/ashuio/shavee) using your project, which is (as far as I can tell) a GPL 3.0 project, although they [might not be aware of this fact](https://github.com/ashuio/shavee/blob/d625a6bc7fd8a98ffac0c24a46adc8b4492e4d02/LICENSE).

Question is, whether you would like to support another / more permissive license now or in future. If not, that is totally fine and your choice, I just wanted to ask.

Background: I am currently looking for an alternative pam library for an [own project](https://rechenknecht.net/mixxplorer/guest-users). Your library seems to be the best pam lib for modules I discovered so far, but I would like to distribute my library under MIT license, which is not compatible with using `pam_sm_rust`.

**Comment by rcatolino** [comment]
Sorry, I rather like the GPL 3 because I very much want to encourage any work that is "based on" (in the GPL sense) this project to be free/open source as well.

> Background: I am currently looking for an alternative pam library for an [own project](https://rechenknecht.net/mixxplorer/guest-users). Your library seems to be the best pam lib for modules I discovered so far, but I would like to distribute my library under MIT license, which is not compatible with using pam_sm_rust.

I'm a bit confused by that. I thought the MIT was compatible with the GPL, in the sense that you are free to license the code of your project under the MIT license, although distributing the project as a whole would still have to comply with the GPL.

**Comment by lmm-git** [comment]
Hi,

thanks for you quick response!

As far as I understand MIT is actually compatible with GPL. That means MIT licensed projects can be used in GPL licensed projects and GPL licensed projects (like libraries) can be used in MIT licensed projects. But there is a catch: In the latter case when both projects are used together (the library gets used) the whole software is licensed as GPL.

Therefore, I could license my software under MIT and others could take parts of it (which are not interacting with a GPL licensed library) and use it under the MIT, but the software as a whole would be licensed under GPL due to the inclusion of a GPL library. And therefore, when another lib is using my software, it would have the same problem again.

Essentially this was the idea of the GPL, to infect every software to have open sourced everything. Although I like the idea, it is currently not really feasible, at least for me. That said, I will probably not start using your library for this reason.

References

* https://opensource.stackexchange.com/a/1641
* https://www.gnu.org/licenses/gpl-faq.html#IfLibraryIsGPL

**Comment by m00nwtchr** [comment]
@rcatolino 
> Sorry, I rather like the GPL 3 because I very much want to encourage any work that is "based on" (in the GPL sense) this project to be free/open source as well.

Consider the [Mozilla Public License](https://en.wikipedia.org/wiki/Mozilla_Public_License), which only requires that modifications to the MPL-licensed code need to be republished under the same license, but not the software that depends on a MPL-licensed library.
